### PR TITLE
Add SQLite support

### PR DESCRIPTION
**Pull request by pierreburgy**
My PR is a:
- [ ] 💥 Breaking change
- [ ] 🐛 Bug fix #issueNumber
- [ ] 💅 Enhancement
- [x] 🚀 New feature

Main update on the:
- [ ] Admin
- [ ] Documentation
- [x] Framework
- [ ] Plugin

SQLite support has been highly demanded (#1562 and https://portal.productboard.com/strapi/c/35-support-sqlite-database). It will make setup much easier. Also, we will be able to offer a nice live demo using CodeSandbox 🎉

Search is currently using LIKE syntax. If necessary, I can add [FTS5](https://sqlite.org/fts5.html) and triggers like explaining in this [great article](https://medium.com/@JasonWyatt/squeezing-performance-from-sqlite-indexes-indexes-c4e175f3c346) written by @jasonwyatt.

### CODE REVIEW

**Comment by derrickmehaffy** [comment]
Nice one!

**Comment by Aurelsicoko** [comment]
Nice job 👍 

However, we should remember why we removed the SQLite support 8 months ago... (see https://github.com/strapi/strapi/pull/1025). It was probably due to a lack of time (cf https://github.com/strapi/strapi/issues/485) but we should be careful about this re-implementation.

**Comment by derrickmehaffy** [comment]
> Also, I was thinking about removing completely the database's step during the project's generation process. 95% of the time SQLite will perfectly match with the needs of the developer.

Honestly that would probably be a good choice, let the users configure the database themselves of have another option with the CLI tool to configure one. I see a lot of questions related to migration process and creating a new app (short term). However I think merging the templates used would be best before this is done. (generate:api command, and admin generation)

**Comment by pierreburgy** [comment]
This PR is now ready for tests 🎉

Note 1: even in --dev mode, generated projects use the `strapi-hook-bookshelf` from npm. The problem is that, obviously, the `connectivity.js` file on npm does not contains the changes made in this PR. So, an error occurs when generating a new project. The work around is to generate a new project with Postgres and to replace the content of `config/environments/development/database.json` with:

```
{
  "defaultConnection": "default",
  "connections": {
    "default": {
      "connector": "strapi-hook-bookshelf",
      "settings": {
        "client": "sqlite",
         "fileName": ".tmp/data.db"
      },
      "options": {
        "useNullAsDefault": true
      }
    }
  }
}
```

Maybe @lauriejim has a work around to test it with the correct `connectivity.js` file.

Note 2: by default, the SQLite file is generated in `.tmp` (which is in the `.gitignore` of generated projects) with the name `data.db`. So the result is `.tmp/data.db`. Feel free to change it if necessary.

**Comment by pierreburgy** [comment]
@mnlbox thank you for your comment!

1. @Aurelsicoko could you move this feature to "In Progress", please?

2. Exactly, this is why import/export feature is highly requested.

3. This is interesting, but I think offering Strapi online would solve this problem in a more straightforward way.

File: packages/strapi-hook-bookshelf/lib/index.js [hunk 1]
```diff
@@ -470,6 +470,7 @@ module.exports = function(strapi) {
                       try {
                         const connection = strapi.config.connections[definition.connection];
                         let columns = Object.keys(attributes).filter(attribute => ['string', 'text'].includes(attributes[attribute].type));
+                        let indexes;
```
Comment (Aurelsicoko) [inline packages/strapi-hook-bookshelf/lib/index.js]: Why? It seems that the variable isn't used anywhere. We should keep it in the scope of the `if` condition?

File: packages/strapi-hook-knex/lib/index.js [hunk 2]
```diff
@@ -137,9 +142,8 @@ module.exports = strapi => {
             break;
           case 'sqlite3':
             // Create the directory if it does not exist.
-            const directory = path.dirname(path.resolve(strapi.config.appPath, options.connection.filename));
-            if (!fs.existsSync(directory)){
-              fs.mkdirSync(directory);
+            if (!fs.existsSync(fileDirectory)){
```
Comment (Aurelsicoko) [inline packages/strapi-hook-knex/lib/index.js]: This function is deprecated (see https://nodejs.org/api/fs.html#fs_fs_exists_path_callback). You should use `fs.access` instead.

Comment (soupette) [reply packages/strapi-hook-knex/lib/index.js]: I know that we often use such functions in the codebase to check whether or not a file or a folder does exactly exist like in here https://github.com/strapi/strapi/blob/39f4d8884f5d963dd00454b04e19f17cce9f2f4d/packages/strapi-plugin-documentation/services/Documentation.js#L35-L40

@Aurelsicoko I think that not only we @pierreburgy should change this function and put it the strapi-utils...

**Comment by lauriejim** [comment]
Relations with plugins model are broken. We have to fix it.

**Pull request merged**